nio/client/async_client.py

```diff
@@ -2302,7 +2331,19 @@ def provider(got_429, got_timeouts):
                 # We have to restart from scratch
                 monitor.transferred = 0
 
-            data = data_provider(got_429, got_timeouts)
+            if isinstance(data_provider, Callable):
+                data = data_provider(got_429, got_timeouts)
+
+            elif isinstance(data_provider, SynchronousFile) or \
+                isinstance(data_provider, AsyncFile):
+                data = data_provider
```

Comment:
When this provider function is called again after we get 429 or timeouts, the file should be seeked back to where it was, else reading it a second time would return nothing. 
You can have something like:
```py3
initial_file_position = 0  # before the `def provider`
```

```py3
if got_429 or got_timeouts:
    await data_provider.seek(initial_file_position)
else:
    nonlocal initial_file_position
    initial_file_position = await data_provider.tell()
```
(of course there should be the equivalent code without `await`s if the file is synchronous)

Reply:
Shoot I just assumed that `.read()` will always try to read from the start of the file but you're totally right that it doesn't. Will update with a fix